### PR TITLE
[No QA][TS migration] Adjust ESLint and TS configs

### PR DESCRIPTION
**Pull request by blazejkustra**
### Fixed Issues
$ https://github.com/Expensify/App/issues/39119
PROPOSAL: N/A

### Tests

- Open the code, try to change the code in multiple places and verify that:
  - It works reliably every time
  - Performance is as usual
  - ESLint and TS catches errors according to rules/options in configs 

### Offline tests
N/A

### QA Steps

N/A

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
### Screenshots/Videos

N/A

### CODE REVIEW

File: .eslintrc.js
Comment (fabioh8010) [inline .eslintrc.js:107]: Do we still need these ones?

File: .github/actions/javascript/authorChecklist/index.js
Comment (fabioh8010) [inline .github/actions/javascript/authorChecklist/index.js:1]: I'm curious, why did these generated files change?

File: web/gtm.js
Comment (BartoszGrajdek) [inline web/gtm.js:1]: AFAIK this code is generated by google tag manager and later pasted here, maybe we should just put it in `.eslintignore`?

File: tsconfig.json
Comment (BartoszGrajdek) [inline tsconfig.json:34]: Are you sure this is something we want to remove? I know that we will be more typesafe without it, but I'm not sure if it's a good call in project with so many dependencies

File: src/components/OptionsSelector/BaseOptionsSelector.js
Comment (BartoszGrajdek) [inline src/components/OptionsSelector/BaseOptionsSelector.js:1]: Just to make sure - This is something temporary, right? We put these comments in files that are going to be refactored? 👀 

Maybe it would make sense to add a comment with link to the GH issue where these lodash functions will be removed?

File: src/pages/iou/request/MoneyTemporaryForRefactorRequestParticipantsSelector.js
Comment (BartoszGrajdek) [inline src/pages/iou/request/MoneyTemporaryForRefactorRequestParticipantsSelector.js:1]: Same here

**Review by BartoszGrajdek — COMMENTED** [review]
Left a few comments 👀

File: tsconfig.json
Comment (blazejkustra) [reply tsconfig.json:34]: I removed this one as we extend expo config which already has this option, so no real changes here

File: src/components/OptionsSelector/BaseOptionsSelector.js
Comment (blazejkustra) [reply src/components/OptionsSelector/BaseOptionsSelector.js:1]: Yes it is temporary, these will be handled in some future refactor PRs.

Comment (blazejkustra) [reply src/components/OptionsSelector/BaseOptionsSelector.js:1]: Good idea overall, I'm going to add additional comment to the code

File: .eslintrc.js
Comment (blazejkustra) [reply .eslintrc.js:107]: I checked it yesterday. When I remove it I get around 1000 new TS errors, I'm going to handle these in a separate PR 😮‍💨

**Review by BartoszGrajdek — APPROVED** [review]
LGTM

**Review by fabioh8010 — APPROVED** [review]

File: .eslintrc.js
Comment (fabioh8010) [reply .eslintrc.js:107]: 🙈

File: .github/actions/javascript/authorChecklist/index.js
Comment (fabioh8010) [reply .github/actions/javascript/authorChecklist/index.js:1]: I guess it's because of the `.eslintrc.js` added?

File: web/gtm.js
Comment (blazejkustra) [reply web/gtm.js:1]: I just moved exclude two lines below 😅 It's for consistency reasons, as it makes more sense to have `include` first

Comment (blazejkustra) [reply web/gtm.js:1]: Ops, it is a mistake on my end. Good catch 👍

**Comment by blazejkustra** [comment]
@parasharrajat Fixed! [TS check is failing](https://github.com/Expensify/App/actions/runs/8829211919/job/24239745230?pr=40778) because JS files were added (.eslintrc.js). So no need to worry about this 😄

**Review by parasharrajat — APPROVED** [review]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.


:ribbon: :eyes: :ribbon:  C+ reviewed

**Comment by mountiny** [comment]
@blazejkustra Why is this on hold now? I see conflicts but not sure why the hold

**Comment by blazejkustra** [comment]
@mountiny I saw some conflicts while I was OOO, so I just added the hold label until I come back. Conflicts are resolved now, all yours! 😄

**Comment by blazejkustra** [comment]
Test Github Actions workflows is failing, I'm investigating

**Comment by blazejkustra** [comment]
All good now @mountiny! I had to [adjust](https://github.com/Expensify/App/pull/40778/commits/ccfd44ba8a5c7b120f499eb8c76df6d3cf48efc0) one of the workflow out of nowhere 🤔 

At first I thought my changes introduced this regression but I started fresh and tested it on another [PR](https://github.com/Expensify/App/pull/41682) with no impactful changes and I got the same [error](https://github.com/Expensify/App/actions/runs/8971806823/job/24638350929?pr=41682).

**Comment by mountiny** [comment]
Sorry, I was ooo, I will merge this after the merge freeze

**Review by roryabraham — COMMENTED** [review]
High-level request: rather than making all these changes here in our `.eslintrc.js`, can we update our base [eslint-config-expensify](https://github.com/Expensify/eslint-config-expensify) to include all the rules (dynamically for JS and TS)?

**Comment by blazejkustra** [comment]
@roryabraham While I agree that it would be nice to move the majority of rules to `eslint-config-expensify`; I don't think it's worth the effort for now. 

This won't be as easy as moving all the rules from E/App `.eslintrc.js` to `eslint-config-expensify`. This would be rather a big refactor of `eslint-config-expensify` with a lot of testing in different repos to make sure TS/React/RN rules are applied to E/App while a different set of rules is used in `react-native-onyx` and the same goes for other repos that Expensify maintains (old mobile/old web/live-markdown/expensify-common etc).

I'm worried that it would take too much time and won't bring much value. 
- How about creating a backlog ticket for this? / Or just do nothing for now. 
- An alternative approach would be to pick all rules that are both used in onyx and E/App and move them to `eslint-config-expensify`.

**Comment by roryabraham** [comment]
I'm confused why it would be a big refactor in eslint-config-expensify, given that most/all the rules here are specific to either TS files or React Native. I'd think that would mean it would have minimal effect on our other repos unless they start using TS or RN.

That said, if you investigated and feel strongly then it's NAB, because as I said I'd expect the effect on other repos to be very limited.

**Comment by roryabraham** [comment]
There are conflicts and failing TS checks here, btw

File: .eslintignore
Comment (roryabraham) [inline .eslintignore:1]: *question:* does this line ignore anything that starts with `.`? If yes, great change! I was going to ask that we ignore `.bundle` and `.expo` too but it seems like that's covered.

File: .eslintrc.js
Comment (roryabraham) [inline .eslintrc.js:103]: Please don't leave TODOs in the code. Create a GitHub issue instead and link it with a comment in the PR

**Review by roryabraham — DISMISSED** [review]
overall this seems good

File: .eslintignore
Comment (blazejkustra) [reply .eslintignore:1]: Previously we had `.eslintignore` file and `ignorePatterns` in eslint config:
```ts
ignorePatterns: ['!.*', 'src/vendor', '.github/actions/**/index.js', 'desktop/dist/*.js', 'dist/*.js', 'node_modules/.bin/**', 'node_modules/.cache/**', '.git/**'],
```

I removed ignorePatterns from the config and unified all patterns. So nothing should really change 👍

File: .eslintrc.js
Comment (blazejkustra) [reply .eslintrc.js:103]: Overall I agree that it is a bad practice leave such TODOs in the codebase. The comment was added at the beginning of TS migration, and will be taken care of as [part of this issue](https://github.com/Expensify/App/issues/39119#issuecomment-2095635912), but in a separate PR.

File: .eslintignore
Comment (blazejkustra) [reply .eslintignore]: Yes! I thought it was clever but turns out here are two caveats:
- With this approach `.github` and `.storybook` weren't linted (it's more complicated because we want to lint `.github` but not `.github/actions/**/index.js`)
- Turns out dot files/directories are automatically omitted so no need for `.*`, instead I did this:
```
!.storybook
!.github
```

Comment (blazejkustra) [reply .eslintignore]: @parasharrajat I checked it myself but could you double check that correct directories are linted (.github, .storybook, src, web, desktop etc) and correct are omitted (.github/actions/**/index.js, configs, dist, .expo, .git)? Thank you!

**Comment by blazejkustra** [comment]
> There are conflicts and failing TS checks here, btw

@roryabraham Left an explanation [here](https://github.com/Expensify/App/pull/40778#issuecomment-2076626584)

**Review by mountiny — DISMISSED** [review]
Looks good to me, the TS failures are expected for now

@roryabraham all yours

**Review by mountiny — APPROVED** [review]

**Comment by blazejkustra** [comment]
> Shouldn't the following comments throw a warning as we are not putting new lines before them? This Expensify.tsx file.

@parasharrajat I believe we don't have a rule for it 🤔

**Comment by blazejkustra** [comment]
> I'm confused why it would be a big refactor in eslint-config-expensify, given that most/all the rules here are specific to either TS files or React Native. I'd think that would mean it would have minimal effect on our other repos unless they start using TS or RN.

You're correct, that most of these rules are indeed specific to TS and RN, but also specific to NewDot. These rules might impact other projects that have different coding styles (Onyx as an example). Such impact might go unnoticed unless thoroughly checked. This leads to the need to restructure and possibly add another preset to `eslint-config-expensify`. Rather than extending from 'expensify' in .eslintrc.js as shown below:  
```ts
extends: [ 'expensify', ...
```

We would have something like this instead:
```ts
extends ['expensify-typescript', ...
```

Breaking it down, this implies:
- Reorganization of the rules, potentially reshaping a significant part of the `eslint-config-expensify` repository
- Collaborative agreement on rules to be applicable across all JS/TS repositories (which could also entail deprecating some JS rules)
- Extensive testing across various repositories, a few of which I don't have access to

That's where the potential time consumption and complexity arises from. @roryabraham Let me your thoughts + please have a final look at the PR 😄

**Review by roryabraham — APPROVED** [review]
LGTM 👍🏼 

TypeCheck is failing because of the addition of a new `.js` file, but it's a `.eslintrc.js`, so not a concern. Going to merge.

**Comment by roryabraham** [comment]
not an emergency - see last comment

**Pull request merged**
